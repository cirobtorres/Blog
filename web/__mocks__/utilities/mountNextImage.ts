const nextOptimizedImage = (filePath: string) => {
  if (filePath) {
    const filename = filePath.substring(
      filePath.lastIndexOf("/") + 1,
      filePath.lastIndexOf(".")
    );
    const extension = filePath.substring(filePath.lastIndexOf(".") + 1);
    const path = filePath.substring(0, filePath.lastIndexOf("/") + 1);
    const URL = `${path}${filename}.${extension}`;
    const encodedURL = encodeURIComponent(URL);
    const width = "w=3840";
    const quality = "q=75";
    const nextOptimizedImage = `/_next/image?url=${encodedURL}&${width}&${quality}`;
    return nextOptimizedImage;
  }
  return null;
};

interface ToEncodeURIAttributes {
  imageFilename: string;
  imageFileExtension: string;
  path: string;
  width: number;
}

function mountLocalImage({
  imageFilename,
  imageFileExtension,
  path,
  width,
}: ToEncodeURIAttributes) {
  // Mock local images
  const quality = 75;
  const URL = `${path}${imageFilename}.${imageFileExtension}`;
  const encodedURL = encodeURIComponent(URL);
  const nextOptimizedImage = `/_next/image?url=${encodedURL}&w=${width}&q=${quality}`;
  return nextOptimizedImage;
}

function requestBackEndImage(
  filePath?: string,
  width: number = 1920,
  options?: {
    PROTOCOL?: string;
    IP?: string;
    PORT?: string;
  }
) {
  // Mock images managed by Strapi
  const quality = 75;

  if (filePath) {
    const filename = filePath.substring(
      filePath.lastIndexOf("/") + 1,
      filePath.lastIndexOf(".")
    );
    const extension = filePath.substring(filePath.lastIndexOf(".") + 1);
    const path = filePath.substring(0, filePath.lastIndexOf("/") + 1);

    const protocol = options ? options.PROTOCOL : "http";
    const ip = options ? options.IP : "127.0.0.1";
    const port = options ? options.PORT : "1337";
    const backendURL = `${protocol}://${ip}:${port}`;

    const URL = `${backendURL}${path}${filename}.${extension}`;

    const encodedURL = encodeURIComponent(URL);
    const nextOptimizedImage = `/_next/image?url=${encodedURL}&w=${
      width * 2
    }&q=${quality}`;

    return nextOptimizedImage;
  }

  const URL = "/images/not-authenticated.png";
  const encodedURL = encodeURIComponent(URL);

  const nextOptimizedImage = `/_next/image?url=${encodedURL}&w=${
    width * 2
  }&q=${quality}`;

  return nextOptimizedImage;
}

export { nextOptimizedImage, mountLocalImage, requestBackEndImage };
