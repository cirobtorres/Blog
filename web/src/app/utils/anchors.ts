const extractAnchors = (htmlString: string): { [key: string]: string }[] => {
  const anchorList: { [key: string]: string }[] = [];
  const regex = /<(h[1-6])([^>]*)>(.*?)<\/\1>/gi;

  let match;
  while ((match = regex.exec(htmlString)) !== null) {
    const [, tag, attributes, content] = match;

    // Skip if content has HTML tags
    if (/<[^>]+>/.test(content)) continue;

    // Get id or generate one
    const idMatch = attributes.match(/id="([^"]*)"/);
    const id = idMatch
      ? idMatch[1]
      : content
          .trim()
          .toLowerCase()
          .replace(/\s+/g, "-")
          .replace(/[^\w\-]/g, "");

    anchorList.push({ [id]: `<${tag}>${content}</${tag}>` });
  }

  return anchorList;
};

const addIdsToHeadings = (htmlString: string) => {
  return htmlString.replace(/<(h[1-6])>(.*?)<\/\1>/g, (match, tag, content) => {
    const id = content
      .trim()
      .toLowerCase()
      .replace(/\s+/g, "-") // replace whitespaces for "-"
      .replace(/[^\w\-]/g, ""); // Remove special characters
    return `<${tag} id="${id}">${content}</${tag}>`;
  });
};

export { addIdsToHeadings, extractAnchors };
