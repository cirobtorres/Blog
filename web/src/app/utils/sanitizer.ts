import sanitizeHtml from "sanitize-html";

const sanitizer = (comment: string) => {
  return sanitizeHtml(comment, {
    allowedTags: ["p"],
    // allowedAttributes: {
    //   a: ["href", "target"],
    // },
  });
};

export default sanitizer;
