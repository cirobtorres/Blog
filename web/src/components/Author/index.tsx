import Link from "next/link";
import Image from "next/image";
import { Skeleton } from "../Shadcnui/skeleton";

const Author = ({ author }: { author: Author }) => {
  return (
    <div data-id="author" className="flex gap-2 items-center">
      <Link href="/sobre" className="group">
        <div className="relative flex size-10 shrink-0 overflow-hidden rounded-full">
          <Image
            src={
              author.avatar
                ? `http://127.0.0.1:1337${author.avatar.url}`
                : "/images/not-authenticated.png"
            }
            alt={
              author.avatar
                ? author.avatar.alternativeText
                : `Avatar de ${author.name}`
            }
            fill
            sizes="(max-width: 40px) 100vw"
            className="transition-all duration-500 absolute object-cover group-hover:brightness-50"
          />
        </div>
      </Link>
      <Link href="/sobre">
        <p className="transition-colors duration-500 hover:text-blog-foreground-readable-hover text-blog-foreground-highlight">
          <strong>@{author.name}</strong>
        </p>
      </Link>
    </div>
  );
};

const AuthorSkeleton = () => {
  return (
    <div className="flex gap-4 items-center">
      <Skeleton className="size-10 shrink-0 rounded-full" />
      <Skeleton className="w-20 h-4 shrink-0 rounded-full" />
    </div>
  );
};

export default Author;
export { AuthorSkeleton };
