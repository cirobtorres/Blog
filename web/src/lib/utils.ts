import { remark } from "remark";
import html from "remark-html";
import { clsx, type ClassValue } from "clsx";
import { twMerge } from "tailwind-merge";

const cn = (...inputs: ClassValue[]) => {
  return twMerge(clsx(inputs));
};

const convertMarkdowToHtmlString = async (rawHtml: string | undefined) => {
  const processedContent = await remark().use(html).process(rawHtml);
  const stringHtml = processedContent.toString();
  return stringHtml;
};

export { cn, convertMarkdowToHtmlString };
