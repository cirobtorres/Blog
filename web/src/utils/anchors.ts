import slugify from "./slugify";

const extractAnchors = (htmlString: string): { [key: string]: string }[] => {
  const anchorList: { [key: string]: string }[] = [];
  const regex = /<(h[1-6])([^>]*)>(.*?)<\/\1>/gi;

  let match;

  while ((match = regex.exec(htmlString)) !== null) {
    const [, tag, attributes, content] = match;

    // Skip if content has HTML tags like <b> <strong> <i> <strike> <u>
    // Ex: <h1>Example <b>1</b></h1>
    if (/<[^>]+>/.test(content)) continue;

    // Get id or generate one
    const idMatch = attributes.match(/id="([^"]*)"/);
    const id = idMatch ? idMatch[1] : slugify(content);

    anchorList.push({ [id]: `<${tag}>${content}</${tag}>` });
  }

  return anchorList;
};

const addIdsToHeadings = (htmlString: string) => {
  return htmlString.replace(/<(h[1-6])>(.*?)<\/\1>/g, (match, tag, content) => {
    const id = slugify(content);
    return `<${tag} id="${id}" tabIndex="0">${content}</${tag}>`;
  });
};

export { addIdsToHeadings, extractAnchors };
