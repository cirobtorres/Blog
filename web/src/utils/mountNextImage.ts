interface ToEncodeURIAttributes {
  imageFilename: string;
  imageFileExtension: string;
  path: string;
  width: number;
  quality: number;
}

function mountNextImage({
  imageFilename,
  imageFileExtension,
  path,
  width,
  quality,
}: ToEncodeURIAttributes) {
  const URL = `${path}${imageFilename}.${imageFileExtension}`;
  const encodedURL = encodeURIComponent(URL);
  const nextOptimizedImage = `/_next/image?url=${encodedURL}&w=${width}&q=${quality}`;
  return nextOptimizedImage;
}

function requestBackEndNextImage(
  imageFilename: string,
  imageFileExtension: string,
  path: string,
  width: number,
  quality: number,
  options?: {
    PROTOCOL?: string;
    IP?: string;
    PORT?: string;
  }
) {
  const URL = `${options ? options.PROTOCOL : "http"}://${
    options ? options.IP : "127.0.0.1"
  }:${
    options ? options.PORT : "1337"
  }${path}${imageFilename}.${imageFileExtension}`;
  const encodedURL = encodeURIComponent(URL);
  const nextOptimizedImage = `/_next/image?url=${encodedURL}&w=${width}&q=${quality}`;

  return nextOptimizedImage;
}

export { mountNextImage, requestBackEndNextImage };
