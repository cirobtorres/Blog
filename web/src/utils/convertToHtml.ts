import { addIdsToHeadings } from "./anchors";
import convertMarkdowToHtmlString from "./markdown";

const convertToHTML = async (blocks: string) => {
  const contentHtml = await convertMarkdowToHtmlString(blocks);
  const htmlToRender = addIdsToHeadings(contentHtml);
  return htmlToRender;
};

export { convertToHTML };
