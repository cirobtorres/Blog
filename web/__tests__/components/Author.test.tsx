import { requestBackEndNextImage } from "@/utils/mountNextImage";
import Author, { AuthorSkeleton } from "../../src/components/Author";
import { render, screen, waitFor } from "@testing-library/react";

describe("Author (has avatar)", () => {
  const authorMock = {
    documentId: "123",
    name: "John Doe",
    avatar: {
      documentId: "1133445",
      url: "/path/to/avatar.jpg",
      alternativeText: "Avatar de John Doe",
    },
  };

  it("renders author's name inside link", () => {
    render(<Author author={authorMock} />);

    const authorLink = screen.getByRole("link", { name: /@John Doe/i });

    expect(authorLink).toBeInTheDocument();
    expect(authorLink).toHaveAttribute("href", "/sobre");
  });

  it("renders author's avatar inside link", async () => {
    render(<Author author={authorMock} />);

    const avatarImage = screen.getByAltText(/Avatar de John Doe/i);

    const nextOptimizedImage = requestBackEndNextImage(
      "avatar",
      "jpg",
      "/path/to/",
      3840,
      75
    );

    await waitFor(() =>
      // next Image loads components lazily
      // Without waitFor, nextOptimizedImage would be null
      // Either use waitFor or set Image in the component as priority
      expect(avatarImage.getAttribute("src")).toBe(nextOptimizedImage)
    );
  });

  it("checks if link points towards /sobre", () => {
    render(<Author author={authorMock} />);

    const avatarLink = screen.getAllByRole("link");

    avatarLink.forEach((link) =>
      expect(link).toHaveAttribute("href", "/sobre")
    );
  });

  it("matches the snapshot", () => {
    const { asFragment } = render(<Author author={authorMock} />);
    expect(asFragment()).toMatchSnapshot();
  });
});

describe("Author (has no avatar)", () => {
  const authorMock = {
    documentId: "12345",
    name: "John Doe",
    avatar: null,
  };

  it("renders author placeholder image", async () => {
    render(<Author author={authorMock} />);
    const authorAvatarImage = screen.getByAltText(
      /Avatar de John Doe/i
    ) as HTMLImageElement;

    const PLACEHOLDER_IMAGE_NAME = "not-authenticated";
    const PLACEHOLDER_IMAGE_FILE_EXTENSION = "png";
    const PATH = "/images/";
    const URL = `${PATH}${PLACEHOLDER_IMAGE_NAME}.${PLACEHOLDER_IMAGE_FILE_EXTENSION}`;
    const encodedURL = encodeURIComponent(URL);
    const width = "w=3840";
    const quality = "q=75";
    const nextOptimizedImage = `/_next/image?url=${encodedURL}&${width}&${quality}`;

    await waitFor(() =>
      // next Image loads components lazily
      // Without waitFor, nextOptimizedImage would be null
      // Either use waitFor or set Image in the component as priority
      expect(authorAvatarImage.getAttribute("src")).toBe(nextOptimizedImage)
    );
  });

  it("renders author image alt text", () => {
    render(<Author author={authorMock} />);
    const authorAvatarImage = screen.getByAltText(/Avatar de John Doe/i);
    expect(authorAvatarImage).toHaveAttribute("alt", "Avatar de John Doe");
  });

  it("matches the snapshot", () => {
    const { asFragment } = render(<Author author={authorMock} />);
    expect(asFragment()).toMatchSnapshot();
  });
});

describe("AuthorSkeleton", () => {
  it("renders successfully", () => {
    const { container } = render(<AuthorSkeleton />);
    expect(container).toBeInTheDocument();
  });

  it("renders each skeleton, one for the avatar and the other one for the username", () => {
    const { container } = render(<AuthorSkeleton />);

    const avatarSkeleton = container.querySelector(".size-10");
    const nameSkeleton = container.querySelector(".w-20");

    expect(avatarSkeleton).toBeInTheDocument();
    expect(nameSkeleton).toBeInTheDocument();
  });
});
