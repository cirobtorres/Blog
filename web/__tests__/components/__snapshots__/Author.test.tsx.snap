// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Author (has avatar) matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex gap-2 items-center"
    data-id="author"
  >
    <a
      class="group"
      href="/sobre"
    >
      <div
        class="relative flex size-10 shrink-0 overflow-hidden rounded-full"
      >
        <img
          alt="Avatar de John Doe"
          class="transition-all duration-500 absolute object-cover group-hover:brightness-50"
          data-nimg="fill"
          decoding="async"
          loading="lazy"
          sizes="(max-width: 40px) 100vw"
          src="/_next/image?url=http%3A%2F%2F127.0.0.1%3A1337%2Fpath%2Fto%2Favatar.jpg&w=3840&q=75"
          srcset="/_next/image?url=http%3A%2F%2F127.0.0.1%3A1337%2Fpath%2Fto%2Favatar.jpg&w=640&q=75 640w, /_next/image?url=http%3A%2F%2F127.0.0.1%3A1337%2Fpath%2Fto%2Favatar.jpg&w=750&q=75 750w, /_next/image?url=http%3A%2F%2F127.0.0.1%3A1337%2Fpath%2Fto%2Favatar.jpg&w=828&q=75 828w, /_next/image?url=http%3A%2F%2F127.0.0.1%3A1337%2Fpath%2Fto%2Favatar.jpg&w=1080&q=75 1080w, /_next/image?url=http%3A%2F%2F127.0.0.1%3A1337%2Fpath%2Fto%2Favatar.jpg&w=1200&q=75 1200w, /_next/image?url=http%3A%2F%2F127.0.0.1%3A1337%2Fpath%2Fto%2Favatar.jpg&w=1920&q=75 1920w, /_next/image?url=http%3A%2F%2F127.0.0.1%3A1337%2Fpath%2Fto%2Favatar.jpg&w=2048&q=75 2048w, /_next/image?url=http%3A%2F%2F127.0.0.1%3A1337%2Fpath%2Fto%2Favatar.jpg&w=3840&q=75 3840w"
          style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
        />
      </div>
    </a>
    <a
      href="/sobre"
    >
      <p
        class="transition-colors duration-500 hover:text-blog-foreground-readable-hover text-blog-foreground-highlight"
      >
        <strong>
          @John Doe
        </strong>
      </p>
    </a>
  </div>
</DocumentFragment>
`;

exports[`Author (has no avatar) matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex gap-2 items-center"
    data-id="author"
  >
    <a
      class="group"
      href="/sobre"
    >
      <div
        class="relative flex size-10 shrink-0 overflow-hidden rounded-full"
      >
        <img
          alt="Avatar de John Doe"
          class="transition-all duration-500 absolute object-cover group-hover:brightness-50"
          data-nimg="fill"
          decoding="async"
          loading="lazy"
          sizes="(max-width: 40px) 100vw"
          src="/_next/image?url=%2Fimages%2Fnot-authenticated.png&w=3840&q=75"
          srcset="/_next/image?url=%2Fimages%2Fnot-authenticated.png&w=640&q=75 640w, /_next/image?url=%2Fimages%2Fnot-authenticated.png&w=750&q=75 750w, /_next/image?url=%2Fimages%2Fnot-authenticated.png&w=828&q=75 828w, /_next/image?url=%2Fimages%2Fnot-authenticated.png&w=1080&q=75 1080w, /_next/image?url=%2Fimages%2Fnot-authenticated.png&w=1200&q=75 1200w, /_next/image?url=%2Fimages%2Fnot-authenticated.png&w=1920&q=75 1920w, /_next/image?url=%2Fimages%2Fnot-authenticated.png&w=2048&q=75 2048w, /_next/image?url=%2Fimages%2Fnot-authenticated.png&w=3840&q=75 3840w"
          style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
        />
      </div>
    </a>
    <a
      href="/sobre"
    >
      <p
        class="transition-colors duration-500 hover:text-blog-foreground-readable-hover text-blog-foreground-highlight"
      >
        <strong>
          @John Doe
        </strong>
      </p>
    </a>
  </div>
</DocumentFragment>
`;
